reassembly/tcpcheck.go

```diff
@@ -0,0 +1,241 @@
+// tree.
+
+package reassembly
+
+import (
+	"encoding/binary"
+	"fmt"
+
+	"github.com/google/gopacket"
+	"github.com/google/gopacket/layers"
+)
+
+/*
+ * Check TCP packet against options (window, MSS)
+ */
+
+type tcpStreamOptions struct {
+	mss           int
+	scale         int
+	receiveWindow uint
+}
+
+// TCPOptionCheck contains options for the two directions
+type TCPOptionCheck struct {
+	options [2]tcpStreamOptions
+}
+
+func (t *TCPOptionCheck) getOptions(dir TCPFlowDirection) *tcpStreamOptions {
+	if dir == TCPDirClientToServer {
+		return &t.options[0]
+	}
+	return &t.options[1]
+}
+
+// NewTCPOptionCheck creates default options
+func NewTCPOptionCheck() TCPOptionCheck {
+	return TCPOptionCheck{
+		options: [2]tcpStreamOptions{
+			tcpStreamOptions{
+				mss:           0,
+				scale:         -1,
+				receiveWindow: 0,
+			}, tcpStreamOptions{
+				mss:           0,
+				scale:         -1,
+				receiveWindow: 0,
+			},
+		},
+	}
+}
+
+// Accept checks whether the packet should be accepted by checking TCP options
+func (t *TCPOptionCheck) Accept(tcp *layers.TCP, ci gopacket.CaptureInfo, dir TCPFlowDirection, acked Sequence, start *bool) error {
+	options := t.getOptions(dir)
+	if tcp.SYN {
+		mss := -1
+		scale := -1
+		for _, o := range tcp.Options {
+			// MSS
+			if o.OptionType == 2 {
+				if len(o.OptionData) != 2 {
+					return fmt.Errorf("MSS option data length expected 2, got %d", len(o.OptionData))
+				}
+				mss = int(binary.BigEndian.Uint16(o.OptionData[:2]))
+			}
+			// Window scaling
+			if o.OptionType == 3 {
+				if len(o.OptionData) != 1 {
+					return fmt.Errorf("Window scaling length expected: 1, got %d", len(o.OptionData))
+				}
+				scale = int(o.OptionData[0])
+			}
+		}
+		options.mss = mss
+		options.scale = scale
+	} else {
+		if acked != invalidSequence {
+			revOptions := t.getOptions(dir.Reverse())
+			length := len(tcp.Payload)
+
+			// Check packet is in the correct window
+			diff := acked.Difference(Sequence(tcp.Seq))
+			if diff == -1 && (length == 1 || length == 0) {
+				// This is probably a Keep-alive
+				// TODO: check byte is ok
+			} else if diff < 0 {
+				return fmt.Errorf("Re-emitted packet (diff:%d,seq:%d,rev-ack:%d)", diff,
+					tcp.Seq, acked)
+			} else if revOptions.mss > 0 && length > revOptions.mss {
+				return fmt.Errorf("%d > mss (%d)", length, revOptions.mss)
+			} else if revOptions.receiveWindow != 0 && revOptions.scale < 0 && diff > int(revOptions.receiveWindow) {
+				return fmt.Errorf("%d > receiveWindow(%d)", diff, revOptions.receiveWindow)
+			}
+		}
+	}
+	// Compute receiveWindow
+	options.receiveWindow = uint(tcp.Window)
+	if options.scale > 0 {
+		options.receiveWindow = options.receiveWindow << (uint(options.scale))
+	}
+	return nil
+}
+
+// TCPSimpleFSM implements a very simple TCP state machine
+//
+// Limitations:
+// - packet should be received in-order.
+// - no check on sequence number is performed
+// - no RST
```

Comment:
Add comment here on when/why a user would want to use this?  I'm unclear as to why I would, and whether it's necessary to export it vs. just keeping it internal to the subpackage.